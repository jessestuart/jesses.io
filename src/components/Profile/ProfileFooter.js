/* @flow */
import React, { Component } from 'react'
import Promise from 'bluebird'
import styled from 'styled-components'
import classNames from 'classnames'
import _ from 'lodash'
import { ChevronDown } from 'react-feather'

import Cancelable from '../../utils/cancelable.js'

const initialState = { isAnimatingChevron: false }

// Configure Bluebird's Promise lib to be cancelable --
// we use this to cancel the bouncing chevron animation
// when the component is unmounted (e.g., because the user
// has navigated away from the Home page).
Promise.config({ cancellation: true })

// $FlowFixMe
export default class ProfileFooter extends Component {
  cancelable: Cancelable<*>
  interval: any
  state: {
    isAnimatingChevron: boolean,
  }

  constructor() {
    super()
    this.state = { ...initialState }
  }

  shouldComponentUpdate() {
    return !this.cancelable.cancelled
  }

  componentDidMount() {
    this.cancelable = new Cancelable(this.toggleChevronAnimation())
  }

  componentWillUnmount() {
    this.cancelChevronAnimation()
  }

  cancelChevronAnimation() {
    clearInterval(this.interval)
    this.cancelable.cancel()
  }

  toggleChevronAnimation() {
    return new Promise(() => {
      this.interval = setInterval(
        () =>
          Promise.delay(2000).then(() => {
            const canceled = _.get(this, 'cancelable.canceled')
            if (!canceled) {
              // $FlowFixMe
              this.setState({
                isAnimatingChevron: !this.state.isAnimatingChevron,
              })
            } else {
              this.cancelChevronAnimation()
            }
          }),
        4000
      )
    })
  }

  render() {
    const { isAnimatingChevron } = this.state
    return (
      <ProfileFooterSection>
        <h4 className="f4 fw7 tc lh-title code">
          I build software<br />
          on the pull requests<br />
          of giants.
        </h4>
        <div className="flex w-100 justify-center mt2">
          <div
            className={classNames({
              'hvr-wobble-vertical hvr-wobble-vertical-animating': isAnimatingChevron,
            })}
          >
            <ChevronDown color="#fc5270" size={'2rem'} />
          </div>
        </div>
      </ProfileFooterSection>
    )
  }
}

const ProfileFooterSection = styled.div.attrs({
  className: 'bb bw2 b--hot-pink bg-purple pt4 pb3 mb1',
})`
  grid-column: 1 / 13;
`
