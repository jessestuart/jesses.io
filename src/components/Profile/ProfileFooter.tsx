import Promise from 'bluebird'
import classNames from 'classnames'
import _ from 'lodash'
import React, { Component } from 'react'
import { ChevronDown } from 'react-feather'

import { Cancelable } from '../../utils/cancelable'
import colors from '../../utils/colors'

// Configure Bluebird's Promise lib to be cancelable -- we use this to cancel
// the bouncing chevron animation when the component is unmounted (e.g.,
// because the user has navigated away from the Home page).
Promise.config({ cancellation: true })

interface State {
  isAnimatingChevron: boolean
}

const initialState = { isAnimatingChevron: false }

export default class ProfileFooter extends Component<{}, State> {
  public readonly state = { ...initialState }
  private cancelable: Cancelable | undefined

  constructor(props: any) {
    super(props)
    this.state = { ...initialState }
  }

  public shouldComponentUpdate() {
    return !_.get(this.cancelable, 'cancelled')
  }

  public componentDidMount() {
    this.cancelable = new Cancelable(this.animateChevron())
  }

  public componentWillUnmount() {
    this.cancelChevronAnimation()
  }

  public render() {
    const { isAnimatingChevron } = this.state
    return (
      <div
        className="b--hot-pink bb bg-purple bw2 mb1 pb3 pt4"
        style={{
          gridColumn: '1 / 13',
        }}
      >
        <h4 className="f4 fira-mono fw4 lh-title tc white-80">
          I build software
          <br />
          on the pull requests
          <br />
          of giants.
        </h4>
        <div className="flex justify-center mt2 w-100">
          <div
            className={classNames({
              'hvr-wobble-vertical hvr-wobble-vertical-animating': isAnimatingChevron,
            })}
          >
            <ChevronDown
              className="mw-5"
              color={colors.primary.main}
              size="30px"
            />
          </div>
        </div>
      </div>
    )
  }

  private animateChevron(): Promise<void> {
    return Promise.delay(1500).then(() => {
      const isAnimationCanceled = _.get(this.cancelable, 'cancelled')
      if (isAnimationCanceled) {
        return this.cancelChevronAnimation()
      }
      return Promise.delay(1500).then(() => {
        if (_.isNil(this.cancelable) || this.cancelable.cancelled) {
          return Promise.resolve()
        }
        this.setState({
          isAnimatingChevron: !this.state.isAnimatingChevron,
        })
        return this.animateChevron()
      })
    })
  }

  private cancelChevronAnimation() {
    this.cancelable && this.cancelable.cancel()
    delete this.cancelable
  }
}
