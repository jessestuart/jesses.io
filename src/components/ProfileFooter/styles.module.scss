@import '../../styles/js-tachyons.scss';
// @import '~tachyons-sass/tachyons.scss';

.gc311 {
  grid-column: 3 / 11;
}

.profileFooterSection {
  grid-column: 1 / 13;
  align-items: center;
  border-bottom: 5px solid $hot-pink;
  border-top: 2px solid $gray-600;
  display: flex;
  font-family: 'monospace';
  flex-direction: column;
  padding: 20px 0 0;
  width: 100%;
}

.chevronWrapper {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 3rem;

  svg {
    width: 2rem;
  }
}
