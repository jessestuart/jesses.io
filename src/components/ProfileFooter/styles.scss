// @import '../../layouts/base.scss';
// @import '~tachyons-sass/tachyons.scss';
@import '../../styles/js-tachyons.scss';

.headerSection {
  align-items: center;
  border-bottom: 5px solid $hot-pink;
  border-top: 2px solid $gray-600;
  display: flex;
  font-family: 'monospace';
  flex-direction: column;
  padding: 20px 0 0;
  width: 100%;
}

.chevronWrapper {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 3rem;

  svg {
    width: 2rem;
  }
}
