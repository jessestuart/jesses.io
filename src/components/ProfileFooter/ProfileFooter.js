import React, { Component } from 'react'
import classNames from 'classnames'

import feather from 'feather-icons'

import styles from './styles.module.scss'

// const initialState = {
//   isAnimatingChevron: false,
// }

let intervals = []

export default class ProfileFooter extends Component {
  // state = { ...initialState }

  constructor() {
    super()
    this.initialState = {
      isAnimatingChevron: false,
    }
    this.state = { ...this.intiialState }
  }

  componentWillMount() {
    try {
      feather.replace({
        color: '#FF3A5C',
        height: '2.5rem',
        width: '2.5rem',
      })
    } catch (e) {}
  }

  componentDidMount() {
    feather.replace({
      color: '#FF3A5C',
      height: '2.5rem',
      width: '2.5rem',
    })
    intervals.forEach(clearInterval)
    intervals = []
    // intervals.push(setInterval(this.animateChevron.bind(this), 4000))
  }

  componentWillUnmount() {
    this.cancelAnimation()
  }

  cancelAnimation() {
    intervals.forEach(interval => clearInterval(interval))
    intervals = []
  }

  animateChevron() {
    this.setState({ isAnimatingChevron: true }, () =>
      setTimeout(() => this.setState({ isAnimatingChevron: false }), 2000)
    )
  }

  render() {
    const { isAnimatingChevron } = this.state
    return (
      <section
        className="bb bw3 b--hot-pink bg-purple pt3 mb1"
        style={{ gridColumn: '1 / 13' }}
      >
        <h4 className="f4 fw7 tc lh-title code">
          I build software<br />
          on the pull requests<br />
          of giants.
        </h4>
        <div
          className={classNames(
            styles.chevronWrapper,
            isAnimatingChevron
              ? 'hvr-wobble-vertical hvr-wobble-vertical-animating'
              : null
          )}
        >
          <i data-feather="chevron-down" />
        </div>
      </section>
    )
  }
}
