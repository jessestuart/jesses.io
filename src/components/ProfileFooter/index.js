import React, { Component } from 'react'
import classNames from 'classnames'

import feather from 'feather-icons'

import './styles.scss'

const initialState = {
  isAnimatingChevron: false,
}

let intervals = []

export default class ProfileDevIcons extends Component {
  state = { ...initialState }

  componentWillMount() {
    try {
      feather.replace({
        color: '#FF3A5C',
        height: '2.5rem',
        width: '2.5rem',
      })
    } catch (e) {}
  }

  componentDidMount() {
    feather.replace({
      color: '#FF3A5C',
      height: '2.5rem',
      width: '2.5rem',
    })
    intervals.forEach(clearInterval)
    intervals = []
    intervals.push(setInterval(this.animateChevron.bind(this), 4000))
  }

  componentWillUnmount() {
    this.cancelAnimation()
  }

  cancelAnimation() {
    intervals.forEach(interval => clearInterval(interval))
    intervals = []
  }

  animateChevron() {
    this.setState({ isAnimatingChevron: true }, () =>
      setTimeout(() => this.setState({ isAnimatingChevron: false }), 2000)
    )
  }

  render() {
    const { isAnimatingChevron } = this.state
    return (
      <section
        className="bg-primary text-white text-center header-section"
        id="features"
      >
        <h4 className="section-heading profile-footer-header">
          I build software<br />
          on the pull requests<br />
          of giants.
        </h4>
        <div
          className={classNames(
            'chevron-wrapper',
            isAnimatingChevron
              ? 'hvr-wobble-vertical hvr-wobble-vertical-animating'
              : null
          )}
        >
          <i data-feather="chevron-down" />
        </div>
      </section>
    )
  }
}
