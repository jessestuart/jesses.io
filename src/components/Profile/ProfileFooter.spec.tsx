import React from 'react'
import renderer from 'react-test-renderer'

import ProfileFooter from './ProfileFooter'

jest.useFakeTimers()

test('Render ProfileFooter component.', () => {
  const component = renderer.create(<ProfileFooter />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render ProfileFooter component after 10s.', () => {
  const component = renderer.create(<ProfileFooter />)

  const tree = component.toJSON()
  jest.runTimersToTime(10000)
  expect(tree).toMatchSnapshot()
})

test('Render ProfileFooter component.', () => {
  const component = renderer.create(<ProfileFooter />)
  component.unmount()
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render ProfileFooter component (cancelled) after 10s.', () => {
  const component = renderer.create(<ProfileFooter />)
  jest.runTimersToTime(5000)
  component.unmount()

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
