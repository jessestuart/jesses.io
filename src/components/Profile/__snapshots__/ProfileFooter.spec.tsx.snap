// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ProfileFooter component (cancelled) after 10s. 1`] = `null`;

exports[`Render ProfileFooter component (cancelled) after 10s. 2`] = `
<div
  className="b--hot-pink bb bg-purple bw2 mb1 pb3 pt4 css-vurnku"
  style={
    Object {
      "gridColumn": "1 / 13",
    }
  }
>
  <div
    className="fw4 lh-title tc white-80 css-seakmb"
  >
    I build software
    <br />
    on the pull requests
    <br />
    of giants.
  </div>
  <div
    className="justify-center mt2 hvr-wobble-vertical hvr-wobble-vertical-animating css-4cffwv"
  >
    <svg
      className="mw-5"
      fill="none"
      height="30px"
      stroke="#ff3a5c"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="30px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`Render ProfileFooter component after 10s. 1`] = `
<div
  className="b--hot-pink bb bg-purple bw2 mb1 pb3 pt4 css-vurnku"
  style={
    Object {
      "gridColumn": "1 / 13",
    }
  }
>
  <div
    className="fw4 lh-title tc white-80 css-seakmb"
  >
    I build software
    <br />
    on the pull requests
    <br />
    of giants.
  </div>
  <div
    className="justify-center mt2 hvr-wobble-vertical hvr-wobble-vertical-animating css-4cffwv"
  >
    <svg
      className="mw-5"
      fill="none"
      height="30px"
      stroke="#ff3a5c"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="30px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`Render ProfileFooter component. 1`] = `
<div
  className="b--hot-pink bb bg-purple bw2 mb1 pb3 pt4 css-vurnku"
  style={
    Object {
      "gridColumn": "1 / 13",
    }
  }
>
  <div
    className="fw4 lh-title tc white-80 css-seakmb"
  >
    I build software
    <br />
    on the pull requests
    <br />
    of giants.
  </div>
  <div
    className="justify-center mt2 css-4cffwv"
  >
    <svg
      className="mw-5"
      fill="none"
      height="30px"
      stroke="#ff3a5c"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="30px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</div>
`;

exports[`Render ProfileFooter component. 2`] = `null`;
